hal/source/gpio_device.cpp

```diff
@@ -0,0 +1,158 @@
+#include "gpio_device.h"
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/algorithm/string/split.hpp>
+
+
+GPIODevice::GPIODevice(NRTFPGADevice & nrt_fpga_device) :
+	nrt_fpga_device_(nrt_fpga_device) {
+	CommonKeyPairs config = {
+	};
+	ReConfig(config);
+}
+
+
+GPIODevice::~GPIODevice() {
+	ClearInitialized();
+}
+
+error_condition GPIODevice::ReConfig(const CommonKeyPairs& config) {
+	ClearInitialized();
+	vector<uint8_t> out_data = {0, 0, 0, 0};
+	nrt_fpga_device_.Set(NRTFPGADevice::kDOData, out_data);
+	SetInitialized();
+
+	return kNoError;
+}
+
+error_condition GPIODevice::Get(const string& key, vector<uint8_t>& result) {
+	if (key.empty())
+		return kInvalidArgument;
+	vector<uint8_t> get_data;
+	if (key == kDIPDataName)
+	{
+		result.resize(2);
```

Comment:
If `nrt_fpga_device_.Get` fails, `result` should be untouched.
```suggestion
		if (nrt_fpga_device_.Get(NRTFPGADevice::kDIData, get_data) != kNoError)
			return kInvalidArgument;
		result.resize(2);
```